### PR TITLE
Feature/gazebo9 DO NOT MERGE

### PR DESCRIPTION
**Pull request by ZacharyTaylor**
Gazebo 9 seems to change half the possible interfaces. This is going to cause a big issue as unless we create some sort of bridge classes or a million preprocessor ifs this will break all old versions.

My vote is we fork the repo and create a "rotors2" with these changes for gazebo9+ as it is the laziest option

### CODE REVIEW

**Comment by mapc-mmmi** [comment]
I have been trying to do soemthing similar for the last couple of days to get a working version for Gazebo 9. I am running Ubuntu 18.04 ROS Melodic, so your sugestion seems like the only vaible solution forward. For some reason I did not notice thice branch of the project.
I will try to check  it out and prehaps add my changes if possible ?
One thing puzzles that me right currently, why are you updating the liftdragplugin from Gazebo manually ?

**Comment by ZacharyTaylor** [comment]
This branch has only existed for about 2 or 3 days so I guess it makes sense that you missed it.

In making the changes I did a lot of find and replace all in all files using sed. Because of this the lift drag plugin was mainly updated at the same time.

I was really just hacking things to get rotors running with minimal effort, feel free to clean up / change anything in this branch.

**Comment by mapc-mmmi** [comment]
Can I push to this branch or should I make my own ?
I am going through the cmake projects one by one and fixing the compile errors.
Currently, I just finished gazebo_bag_plugin.cpp !

**Comment by ZacharyTaylor** [comment]
This branch is building and running on my machine as is, what sort of errors are you getting?

**Comment by mapc-mmmi** [comment]
One example is that, the cmake script wants find octomap_ros that currently does not exist under ROS melodic.

```
CMake Error at /opt/ros/melodic/share/catkin/cmake/catkinConfig.cmake:83 (find_package):
  Could not find a package configuration file provided by "octomap_ros" with
  any of the following names:

    octomap_rosConfig.cmake
    octomap_ros-config.cmake

  Add the installation prefix of "octomap_ros" to CMAKE_PREFIX_PATH or set
  "octomap_ros_DIR" to a directory containing one of the above files.  If
  "octomap_ros" provides a separate development package or SDK, be sure it
  has been installed.
Call Stack (most recent call first):
  CMakeLists.txt:96 (find_package)


make: *** [cmake_check_build_system] Error 1

```

**Comment by mapc-mmmi** [comment]
one of the changes I made in the cmake script was:
if (NOT NO_ROS AND BUILD_OCTOMAP_PLUGIN)
  find_package(catkin REQUIRED COMPONENTS
    cmake_modules
    cv_bridge
    geometry_msgs
    mav_msgs
    octomap_msgs
    octomap_ros
    rosbag
    roscpp
    rotors_comm
    rotors_control
    std_srvs
    tf
  )
elseif (NOT NO_ROS AND NOT BUILD_OCTOMAP_PLUGIN)
  find_package(catkin REQUIRED COMPONENTS
    cmake_modules
    cv_bridge
    geometry_msgs
    mav_msgs
    rosbag
    roscpp
    rotors_comm
    rotors_control
    std_srvs
    tf
  )
endif()

**Comment by mapc-mmmi** [comment]
Another example for the compiler is this:

```
gazebo_bag_plugin.cpp:133:26: error: ‘class gazebo::physics::World’ has no member named ‘PhysicsEngine’; did you mean ‘PhysicsEnabled’?
   contact_mgr_ = world_->PhysicsEngine()->GetContactManager();
                          ^~~~~~~~~~~~~
                          PhysicsEnabled
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp: In member function ‘void gazebo::GazeboBagPlugin::OnUpdate(const gazebo::common::UpdateInfo&)’:
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:160:30: error: ‘class gazebo::physics::World’ has no member named ‘GetSimTime’; did you mean ‘SetSimTime’?
   common::Time now = world_->GetSimTime();
                              ^~~~~~~~~~
                              SetSimTime
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp: In member function ‘void gazebo::GazeboBagPlugin::ImuCallback(const ImuConstPtr&)’:
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:259:30: error: ‘class gazebo::physics::World’ has no member named ‘GetSimTime’; did you mean ‘SetSimTime’?
   common::Time now = world_->GetSimTime();
                              ^~~~~~~~~~
                              SetSimTime
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp: In member function ‘void gazebo::GazeboBagPlugin::ExternalForceCallback(const WrenchStampedConstPtr&)’:
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:266:30: error: ‘class gazebo::physics::World’ has no member named ‘GetSimTime’; did you mean ‘SetSimTime’?
   common::Time now = world_->GetSimTime();
                              ^~~~~~~~~~
                              SetSimTime
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp: In member function ‘void gazebo::GazeboBagPlugin::WaypointCallback(const MultiDOFJointTrajectoryConstPtr&)’:
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:273:30: error: ‘class gazebo::physics::World’ has no member named ‘GetSimTime’; did you mean ‘SetSimTime’?
   common::Time now = world_->GetSimTime();
                              ^~~~~~~~~~
                              SetSimTime
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp: In member function ‘void gazebo::GazeboBagPlugin::CommandPoseCallback(const PoseStampedConstPtr&)’:
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:280:30: error: ‘class gazebo::physics::World’ has no member named ‘GetSimTime’; did you mean ‘SetSimTime’?
   common::Time now = world_->GetSimTime();
                              ^~~~~~~~~~
                              SetSimTime
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp: In member function ‘void gazebo::GazeboBagPlugin::AttitudeThrustCallback(const AttitudeThrustConstPtr&)’:
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:287:30: error: ‘class gazebo::physics::World’ has no member named ‘GetSimTime’; did you mean ‘SetSimTime’?
   common::Time now = world_->GetSimTime();
                              ^~~~~~~~~~
                              SetSimTime
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp: In member function ‘void gazebo::GazeboBagPlugin::ActuatorsCallback(const ActuatorsConstPtr&)’:
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:295:30: error: ‘class gazebo::physics::World’ has no member named ‘GetSimTime’; did you mean ‘SetSimTime’?
   common::Time now = world_->GetSimTime();
                              ^~~~~~~~~~
                              SetSimTime
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp: In member function ‘void gazebo::GazeboBagPlugin::RateThrustCallback(const RateThrustConstPtr&)’:
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:302:30: error: ‘class gazebo::physics::World’ has no member named ‘GetSimTime’; did you mean ‘SetSimTime’?
   common::Time now = world_->GetSimTime();
                              ^~~~~~~~~~
                              SetSimTime
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp: In member function ‘void gazebo::GazeboBagPlugin::WindSpeedCallback(const WindSpeedConstPtr&)’:
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:309:30: error: ‘class gazebo::physics::World’ has no member named ‘GetSimTime’; did you mean ‘SetSimTime’?
   common::Time now = world_->GetSimTime();
                              ^~~~~~~~~~
                              SetSimTime
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp: In member function ‘void gazebo::GazeboBagPlugin::LogGroundTruth(gazebo::common::Time)’:
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:339:3: error: ‘math’ has not been declared
   math::Pose pose = link_->GetWorldPose();
   ^~~~
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:343:30: error: ‘pose’ was not declared in this scope
   pose_msg.pose.position.x = pose.pos.x;
                              ^~~~
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:343:30: note: suggested alternative: ‘powh’
   pose_msg.pose.position.x = pose.pos.x;
                              ^~~~
                              powh
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:354:54: error: ‘class gazebo::physics::Link’ has no member named ‘GetWorldLinearVel’; did you mean ‘WorldLinearVel’?
   ignition::math::Vector3d  linear_veloctiy = link_->GetWorldLinearVel();
                                                      ^~~~~~~~~~~~~~~~~
                                                      WorldLinearVel
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:355:55: error: ‘class gazebo::physics::Link’ has no member named ‘GetWorldAngularVel’; did you mean ‘WorldAngularVel’?
   ignition::math::Vector3d  angular_veloctiy = link_->GetWorldAngularVel();
                                                       ^~~~~~~~~~~~~~~~~~
                                                       WorldAngularVel
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:359:46: error: ‘ignition::math::v4::Vector3d {aka class ignition::math::v4::Vector3<double>}’ has no member named ‘x’
   twist_msg.twist.linear.x = linear_veloctiy.x;
                                              ^
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:360:46: error: ‘ignition::math::v4::Vector3d {aka class ignition::math::v4::Vector3<double>}’ has no member named ‘y’
   twist_msg.twist.linear.y = linear_veloctiy.y;
                                              ^
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:361:46: error: ‘ignition::math::v4::Vector3d {aka class ignition::math::v4::Vector3<double>}’ has no member named ‘z’
   twist_msg.twist.linear.z = linear_veloctiy.z;
                                              ^
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:362:48: error: ‘ignition::math::v4::Vector3d {aka class ignition::math::v4::Vector3<double>}’ has no member named ‘x’
   twist_msg.twist.angular.x = angular_veloctiy.x;
                                                ^
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:363:48: error: ‘ignition::math::v4::Vector3d {aka class ignition::math::v4::Vector3<double>}’ has no member named ‘y’
   twist_msg.twist.angular.y = angular_veloctiy.y;
                                                ^
/home/martin/ros_sim_drone/src/rotors_simulator/rotors_gazebo_plugins/src/gazebo_bag_plugin.cpp:364:48: error: ‘ignition::math::v4::Vector3d {aka class ignition::math::v4::Vector3<double>}’ has no member named ‘z’
   twist_msg.twist.angular.z = angular_veloctiy.z;
```
These lines also needs to be updated for gazebo9.
`common::Time now = world_->GetSimTime();`
should be changed to:
`common::Time now = world_->SimTime();`

**Comment by ZacharyTaylor** [comment]
While octomap_ros does not have an apt install for melodic you can just clone it into your workspace and build it there https://github.com/OctoMap/octomap_ros

I just realized I forgot to push the final changes I did with the bag plugin, I just added them now.

**Comment by mapc-mmmi** [comment]
Thanks uploading the last changes :-)
I think I will keep my local version of the cmake script, since I am not planning to use the octomap functionality.

File: rotors_gazebo_plugins/include/liftdrag_plugin/liftdrag_plugin.h [hunk 1]
```diff
@@ -107,19 +107,19 @@ namespace gazebo
     protected: double alpha;
 
     /// \brief center of pressure in link local coordinates
-    protected: math::Vector3 cp;
+    protected: ignition::math::Vector3d  cp;
```
Comment (ffurrer) [inline rotors_gazebo_plugins/include/liftdrag_plugin/liftdrag_plugin.h]: remove extra space here and everywhere else.

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/external/gazebo_gimbal_controller_plugin.h [hunk 2]
```diff
@@ -108,9 +108,9 @@ namespace gazebo
     private: common::PID yawPid;
     private: common::Time lastUpdateTime;
 
-    private: ignition::math::Vector3d ThreeAxisRot(
+    private: ignition::ignition::math::Vector3d d ThreeAxisRot(
```
Comment (ffurrer) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/external/gazebo_gimbal_controller_plugin.h]: what are these `d`s for after Vector3d, here and in other places?

**Comment by ZacharyTaylor** [comment]
Just having two copies of all the files and choosing which to use based on gazebo version would be pretty easy to do. However, it would make building new stuff on top really annoying as you would have to copy paste all your changes everywhere.

**Comment by ffurrer** [comment]
test this please

**Comment by ZacharyTaylor** [comment]
I didn't preserve comparability for any gazebo before 9 in this branch so it will never pass jenkins. I think the only clean way to keep compatibility would be to add in a bit of an abstraction layer, though this might be quite a bit of work to do well.

**Comment by michaelpantic** [comment]
@ZacharyTaylor Found an issue with plugins on 18.04/w protobuf and gazebo ( I don't think it was an issue before).

Packages that depend on rotors_gazebo_plugins were looking for threading libraries but couldn't resolve them (-lpthreads). This is connected to 3rd party dependencies that get carried over from importing gazebo. See https://bitbucket.org/osrf/gazebo/issues/2198/gazebo7-linking-error-with-cmake-catkin.

Fix is easy: have rotors_gazebo_plugins depend on gazebo_dev instead of gazebo.

Pushed fix to this branch.

**Comment by antbono** [comment]
Could you post here a detailed procedure to how to install this branch feature/gazebo9 fron scratch? I found a lot of problems compliling hil_interface and mavros extras. I am using ros-kinetic and gazebo9 of course.
Thank you for your attention

**Comment by michaelpantic** [comment]
Fixed with the autobackport script a while ago. closing

**Pull request closed without merging**